cloud/validations.go

```diff
@@ -0,0 +1,256 @@
+	}
+
+	return keyList
+}
+
+// collectTags returns a set of keys for a specified struct tag. If no tag is
+// specified for a particular field of the argument struct type, then the
+// all-lowercase field name is used as per Go tag conventions. If the tag
+// specified is not the name a conventionally formatted go struct tag, then the
+// results of this function are invalid.
+
+func collectTags(t reflect.Type, tag string, ignoreTypes []string, keys *map[string]struct{}) {
+	switch t.Kind() {
+
+	case reflect.Invalid:
+		panic("invalid kind encountered")
+
+	case reflect.Array, reflect.Slice, reflect.Map, reflect.Ptr:
+		collectTags(t.Elem(), tag, ignoreTypes, keys)
+
+	case reflect.Struct:
+		for i := 0; i < t.NumField(); i++ {
+			field := t.Field(i)
+
+			fieldTag := field.Tag.Get(tag)
+			var fieldTagKey string
+
+			ignoredType := false
+			for _, it := range ignoreTypes {
+				if field.Type.String() == it {
+					ignoredType = true
+					break
+				}
+			}
+
+			if fieldTag == "-" || ignoredType {
+				continue
+			}
+
+			if len(fieldTag) > 0 {
+				fieldTagKey = strings.Split(fieldTag, ",")[0]
+			} else {
+				fieldTagKey = strings.ToLower(field.Name)
+			}
+
+			(*keys)[fieldTagKey] = struct{}{}
+			collectTags(field.Type, tag, ignoreTypes, keys)
+		}
+	}
+}
+
+func validateCloudSetMetaData(data []byte) map[string]string {
+	var body interface{}
+	if err := yaml.Unmarshal(data, &body); err != nil {
+		panic(err)
+	}
+
+	jsonBody := yamlToJson(body)
+
+	documentLoader := gojsonschema.NewGoLoader(jsonBody)
+	schemaLoader := gojsonschema.NewGoLoader(cloudSetSchema)
+
+	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
+
+	if err != nil {
+		panic(err)
+	}
+
+	minEditingDistance := 4
+
+	validCloudProperties := CloudTags()
+	suggestionMap := map[string]string{}
+	for _, rsltErr := range result.Errors() {
+		invalidProperty := strings.Split(rsltErr.Description, " ")[2]
+		suggestionMap[invalidProperty] = ""
+		editingDistance := minEditingDistance
+
+		for _, validProperty := range validCloudProperties {
+
+			dist := Distance(invalidProperty, validProperty)
+			if dist < editingDistance && dist < minEditingDistance {
+				editingDistance = dist
+				suggestionMap[invalidProperty] = validProperty
+			}
+
+		}
+	}
+
+	return suggestionMap
+}
+
+func yamlToJson(i interface{}) interface{} {
+	switch x := i.(type) {
+	case map[interface{}]interface{}:
+		m2 := map[string]interface{}{}
+		for k, v := range x {
+			m2[k.(string)] = yamlToJson(v)
+		}
+		return m2
+	case []interface{}:
+		for i, v := range x {
+			x[i] = yamlToJson(v)
+		}
+	}
+	return i
+}
+
+// The following "editing distance" comparator was lifted from
+// https://github.com/gyuho/goling/blob/master/similar/similar.go which has a
+// compatible BSD license. We use it to calculate the distance between a
+// discovered invalid yaml property and known good properties to identify
+// suggestions.
+
+func Distance(str1, str2 string) int {
```

Comment:
this should either go in an internal package (so it's not part of the API), or (preferably) some other utility-type package for other packages to use.